subprojects/execution/src/main/java/org/gradle/internal/execution/caching/CachingState.java

```diff
@@ -18,45 +18,113 @@
 
 import com.google.common.collect.ImmutableList;
 import org.gradle.caching.BuildCacheKey;
+import org.gradle.internal.Either;
+import org.gradle.internal.execution.history.BeforeExecutionState;
 
+import javax.annotation.Nullable;
 import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
 
-public interface CachingState {
-    /**
-     * The cache key if a valid cache key could be built. Might be present even when caching is disabled.
-     */
-    Optional<BuildCacheKey> getKey();
+public class CachingState {
+    public static final CachingState NOT_DETERMINED = disabledWithoutInputs(new CachingDisabledReason(CachingDisabledReasonCategory.UNKNOWN, "Cacheability was not determined"));
+
+    private final Either<Enabled, Disabled> delegate;
```

Comment:
I don't understand why you use a delegate here. You could use two subclasses of `CachingState`, each one having the right things and then implementing the methods.

Reply:
Yeah, we could do that, too. It was quite a bit of back-and-forth, and I landed on this implementation. The upside is that this way the code I use is already tested, so we can rule out some stupid mistakes.

I think a better solution would be to have `EitherBase` that would allow you to expose whatever methods you wanted under whatever names. There's clearly value in using the pattern, but not so sure about how/what to share between the different implementations.